### DISCUSSION TITLE
[RFC] 051 - 知识库/文件上传

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#3023] [RFCs]
## 背景

文件上传应该是目前呼声最高的特性，本 RFC 着手开始研究。

## 思路

首先 LobeChat 应该具有的是「知识库」功能，文件上传是知识库的一个特性。如果从知识库角度出发，它应该具备的功能：

- 基础文件的管理（增删查，暂不考虑修改）；
- 文件详情查阅（支持查看不同类型的文件，允许打开、下载、播放等）
- 添加标签、分组、对文件重命名；
- 支持创建知识库

从 AI 对话的角度来说，需要一个 RAG 能力，那么对应到知识库层面，需要的功能有：
- 文件处理（文件分 Chunk & chunk 向量化），需要关注文件的处理状态
- chunk、embedding 的可视化审查（需要允许合并、删除）
- 文件 chunk 的检索查询（文件内的搜索，chunk 分割测试）
- 打开文件后直接与文件对话

此外，更加进阶的功能包括：
- 导入 Github/GitLab/Notion/Sitemap/Figma/Google Drive 等外部数据到知识库；
- GraphRAG: https://github.com/microsoft/graphrag 

## 功能规划

整个文件上传/知识库功能不太可能一个迭代完成所有功能点，因此考虑将其拆分为多个迭代：

第一期：文件上传基础框架与知识库 : 文件上传独立页面、文件分块与向量化、chunks 预览与语义检索、对话中上传文件与对话，文件格式范围限定为 pdf、txt、word、excel 和 ppt 这五种常见格式，支持文件预览(基于 Portal)，创建知识库。

第二期：能力泛化 - 支持更多种类型的文件分块与向量化与预览、文件夹上传、文件快捷摘要。 Portal 文件预览中支持滑词快捷对话。归集文件 - 整理文件/重命名/移动路径等。

第三期：知识库能力扩展 - 知识库上传市场/ 知识库概览页、 embeddings 聚类分析与 GraphRAG

## 实现路径

- 一期：https://github.com/lobehub/lobe-chat/pull/3487

**Comment by stephonye** [comment]
关于文档上传，也请考虑可以将上传的文档转换格式存储到Vector Database。

**Reply by arvinxx** [reply]
已经考虑到了的

**Comment by arvinxx** [comment]
## 文件上传基础框架

根据几个功能模块列出当前进度

- 文件列表页
  - [x] 列表展示
  - [x] 列表 item 项操作 （下载、复制链接、删除）
  - [x] 空状态
  - [x] 上传按钮
    - [x] 上传文件
    - [x] 上传文件夹
  - [x] 列表查询接口完善
    - [x] 搜索
    - [x] 时间排序、大小排序
    - [x] 文件侧边分类
- 文件上传 Dock
  - [x] 拖拽上传弹窗
  - [x] Dock 组件
  - [x] 文件上传进度与状态
- [x] 文件详情 （待优化）
  - [x] 文件内容预览
    - [x] Text 预览
    - [x] PDF 预览
    - [x]  Word、Excel、PPT 预览
  - [x] 文件详情面板（元信息，未来增加对话模式）
  - [x] 详情 Header（返回、文件名、操作区）

-----

- ✅ 文件处理
  - [x] 文件切块（执行中、出错、成功）
  - [x] chunk 向量化（执行中、成功）

- Chunk 分析
  - [x] 富媒体文件（PDF） chunk 坐标分析
  - [x] 语义化检索（向量化相似度匹配）
    - [x] 有必要增加一层 cache 来降低 query 向量化成本，提升速度

-----

- 知识库
  - [x] 知识库 db schema 设计
  - [x] 知识库 api 路由基础功能（CRUD）
  - [x] 知识库创建、删除、修改名称
  - [x] 知识库侧边栏列表
    - [x] 空状态
  - [x] 知识库详情页
    - [x] 文件列表
    - [ ] 召回测试（二期）
    - [ ] 设置（二期）

------

- 对话文件上传
  - [x] 文件上传模块样式重构
  - [x] 上传过程中的自动处理进度
  - [x] 对话中检索文件向量
    - [x] 粗暴的每次都向量化一遍
    - [x] 缓存到 message 中

- 允许选择知识库/文件范围
  - [x] 允许选中当前会话中文件
  - [x] 允许勾选/取消勾选知识库
  
- 对话中文件的查看
  - [x] 点击文件直接在右侧展示文件详情
  - [x] 返回则进入 Portal 中

- 回答结果的索引与内容直接定位
  - [x] 召回消息缓存到 Assistant Message，并允许查看文件详情

**Reply by arvinxx** [reply]
- [x] chunk/embedding 进度展示是否可做？
- [x] chat 上传走 fast partition 是否合适？fast 完毕以后再异步跑 auto ？
- [x] message query chunks 存储 assistant 消息的关联 chunks
  - [x] 关联 chunks 类型为 Table 或 CompositeElement 。
- [ ] 点击 chunk 直接打开文件，并定位到这个区域，然后高亮这块（对话中文件的查看），需要查到 CompositeElement 所对应的子元素
- [ ] Chunk 任务抛错重试 / Embedding 任务抛错重试
- [x] 文件同 hash 检索，避免重复上传（`file_fingerprint`），
- [ ] 支持文件粒度的删除 chunk 和 embedding
- [x] `messageRAG`  表的必要性？

**Reply by zhangleijun111** [reply]
> * [ ] chunk/embedding 进度展示是否可做？
> * [ ] chat 上传走 fast partition 是否合适？fast 完毕以后再异步跑 auto ？
> * [ ] message RAG Chunks 存储 assistant 消息的关联 chunks
>   * [ ] 关联 chunks 类型为 Table 或 CompositeElement 。
> * [ ] 点击 chunk 直接打开文件，并定位到这个区域，然后高亮这块（对话中文件的查看），需要查到 CompositeElement 所对应的子元素
> * [ ] Chunk 任务抛错重试 / Embedding 任务抛错重试
> * [ ] 文件同 hash 检索，避免重复上传（），`file_fingerprint`
> * [ ] 支持文件粒度的删除 chunk 和 embedding
> * [ ]  `messageRAG`表的必要性？

看来快要成功了

**Reply by BrandonStudio** [reply]
向量会怎么存？服务端在Postgres？客户端呢？

**Reply by arvinxx** [reply]
> 向量会怎么存？服务端在Postgres？客户端呢？

要求 pg 装 vector 插件，然后存在 postgres 里。client db 版本目前暂不计划支持文件上传/知识库相关功能。等 Q4 pglite 方案改造完毕后再看看是否能在 client 文件会话（目前看大概率是不行的）

**Reply by zhangleijun111** [reply]
> > 向量会怎么存？服务端在Postgres？客户端呢？
> 
> 要求 pg 装 vector 插件，然后存在 postgres 里。client db 版本目前暂不计划支持文件上传/知识库相关功能。等 Q4 pglite 方案改造完毕后再看看是否能在 client 文件会话（目前看大概率是不行的）

是要上线了嘛

**Reply by danisvaliev001** [reply]
> > 向量会怎么存？服务端在Postgres？客户端呢？
> 
> 要求 pg 装 vector 插件，然后存在 postgres 里。client db 版本目前暂不计划支持文件上传/知识库相关功能。等 Q4 pglite 方案改造完毕后再看看是否能在 client 文件会话（目前看大概率是不行的）

Is it not that [duckdb](https://github.com/duckdb/duckdb) and [duckdb-wasm](https://github.com/duckdb/duckdb-wasm) looks like more suitable solution than pg/pglite?

**Reply by arvinxx** [reply]
@danisvaliev001 Can duckdb or duckdb-wasm use drizzle as its ORM driver?

**Reply by danisvaliev001** [reply]
> @danisvaliev001 Can duckdb or duckdb-wasm use drizzle as its ORM driver?

Unfortunately, [not yet](https://github.com/drizzle-team/drizzle-orm/discussions/1470).

**Comment by arvinxx** [comment]
核心外部依赖：

- Unstructured.IO: https://docs.unstructured.io/welcome
- gotenberg: https://gotenberg.dev

**Comment by arvinxx** [comment]
一期遗留，二期实现：

- 文件列表分页加载，无限滚动
-  chunk 分页查询 & 无限滚动
-  PPT、EXCEL、WORD chunk 坐标分析 （PPT、EXCEL、Word 转 PDF），仅 Cloud 可用
- Reranker 重排，检索模块性能优化（测了 20+ 用例，目前能有 70% ~ 80% 的成功率）
-  修改知识库图标（后续补充，结合助手图标一同考虑）
-  文件侧边栏知识库排序
-  知识库列表页
-  知识库设置页
- 知识库召回测试页（对话）
- 允许查询所有文件/所有知识库
- 允许选择召回范围（TopK） 和相似度匹配阈值

**Reply by zhangleijun111** [reply]
可以用啦

**Reply by SmallAi-API** [reply]
上线了么

**Comment by arvinxx** [comment]
RAG 相关表结构梳理

<img width="1067" alt="image" src="https://github.com/user-attachments/assets/e2349f33-d6ff-4a84-a1fb-d481b28af6d9">

## chunk 相关

```ts
import { integer, jsonb, pgTable, text, uuid, varchar, vector } from 'drizzle-orm/pg-core';

import { createdAt, updatedAt } from './_helpers';
import { files } from './file';
import { users } from './user';

export const chunks = pgTable('chunks', {
  id: uuid('id').defaultRandom().primaryKey(),
  text: text('text'),
  metadata: jsonb('metadata'),
  index: integer('index'),
  type: varchar('type'),

  createdAt: createdAt(),
  updatedAt: updatedAt(),

  userId: text('user_id').references(() => users.id, { onDelete: 'cascade' }),
  fileId: varchar('file_id').references(() => files.id, { onDelete: 'cascade' }),
});

export type NewChunkItem = typeof chunks.$inferInsert;

export const unstructuredChunks = pgTable('unstructured_chunks', {
  id: uuid('id').defaultRandom().primaryKey(),
  text: text('text'),
  metadata: jsonb('metadata'),
  index: integer('index'),
  type: varchar('type'),

  createdAt: createdAt(),
  updatedAt: updatedAt(),

  parentId: varchar('parent_id'),
  compositeId: uuid('composite_id').references(() => chunks.id, { onDelete: 'cascade' }),
  userId: text('user_id').references(() => users.id, { onDelete: 'cascade' }),
  fileId: varchar('file_id').references(() => files.id, { onDelete: 'cascade' }),
});

export type NewUnstructuredChunkItem = typeof unstructuredChunks.$inferInsert;

export const embeddings = pgTable('embeddings', {
  id: uuid('id').defaultRandom().primaryKey(),
  chunkId: uuid('chunk_id').references(() => chunks.id, { onDelete: 'cascade' }),
  embeddings: vector('embeddings', { dimensions: 1024 }),
  model: text('model'),
  userId: text('user_id').references(() => users.id, { onDelete: 'cascade' }),
});

export type NewEmbeddingsItem = typeof embeddings.$inferInsert;
export type EmbeddingsSelectItem = typeof embeddings.$inferSelect;
```

1. `chunks` 表

这个表用于存储文档的主要分块信息。

- `id`: UUID 类型，作为主键，自动生成
- `text`: 存储分块的具体文本内容
- `metadata`: JSON 类型，用于存储与分块相关的额外元数据
- `index`: 整数类型，用于标记分块在文档中的顺序
- `type`: 字符串类型，用于标识分块的类型（如段落、标题等）
- `createdAt` 和 `updatedAt`: 记录分块的创建和最后更新时间
- `userId`: 外键，链接到 users 表，标识创建该分块的用户
- `fileId`: 外键，链接到 files 表，表示该分块所属的原始文件

2. `unstructuredChunks` 表

这个表专门用于存储使用 unstructured.io 产品进行分块后的结果数据。

- 大部分字段与 `chunks` 表相同，保持了数据结构的一致性
- `parentId`: 用于建立分块之间的层级关系，可能表示更细粒度的分块结构
- `compositeId`: 外键，关联到 chunks 表，明确指出这个 unstructured 分块对应的聚合块

3. `embeddings` 表

这个表用于存储文本嵌入（text embeddings）信息，是实现向量检索的关键。

- `id`: UUID 类型，作为主键，自动生成
- `chunkId`: 外键，关联到 chunks 表，明确指出这个嵌入向量对应的文本分块
- `embeddings`: 向量类型，维度为 1024，存储由文本生成的嵌入向量
- `model`: 文本类型，记录生成嵌入向量所使用的模型
- `userId`: 外键，关联到 users 表，标识创建该嵌入的用户

这些表构成了一个完整的 RAG（检索增强生成）系统的数据基础。`chunks` 表存储基本的文档分块，`unstructuredChunks` 表存储使用 unstructured.io 处理后的更细粒度分块，而 `embeddings` 表则存储这些文本的向量表示。

**Reply by arvinxx** [reply]
## Agent 表

```ts
import { boolean, jsonb, pgTable, primaryKey, text, varchar } from 'drizzle-orm/pg-core';
import { createInsertSchema } from 'drizzle-zod';

import { LobeAgentChatConfig, LobeAgentTTSConfig } from '@/types/agent';

import { idGenerator, randomSlug } from '../../utils/idGenerator';
import { createdAt, updatedAt } from './_helpers';
import { files, knowledgeBases } from './file';
import { users } from './user';

// Agent table is the main table for storing agents
// agent is a model that represents the assistant that is created by the user
// agent can have its own knowledge base and files

export const agents = pgTable('agents', {
  id: text('id')
    .primaryKey()
    .$defaultFn(() => idGenerator('agents'))
    .notNull(),
  slug: varchar('slug', { length: 100 })
    .$defaultFn(() => randomSlug())
    .unique(),
  title: text('title'),
  description: text('description'),
  tags: jsonb('tags').$type<string[]>().default([]),
  avatar: text('avatar'),
  backgroundColor: text('background_color'),

  plugins: jsonb('plugins').$type<string[]>().default([]),
  userId: text('user_id')
    .references(() => users.id, { onDelete: 'cascade' })
    .notNull(),

  chatConfig: jsonb('chat_config').$type<LobeAgentChatConfig>(),

  fewShots: jsonb('few_shots'),
  model: text('model'),
  params: jsonb('params').default({}),
  provider: text('provider'),
  systemRole: text('system_role'),
  tts: jsonb('tts').$type<LobeAgentTTSConfig>(),

  createdAt: createdAt(),
  updatedAt: updatedAt(),
});

export const insertAgentSchema = createInsertSchema(agents);

export type NewAgent = typeof agents.$inferInsert;
export type AgentItem = typeof agents.$inferSelect;

export const agentsKnowledgeBases = pgTable(
  'agents_knowledge_bases',
  {
    agentId: text('agent_id')
      .references(() => agents.id, { onDelete: 'cascade' })
      .notNull(),
    knowledgeBaseId: text('knowledge_base_id')
      .references(() => knowledgeBases.id, { onDelete: 'cascade' })
      .notNull(),
    userId: text('user_id')
      .references(() => users.id, { onDelete: 'cascade' })
      .notNull(),
    enabled: boolean('enabled').default(true),
    createdAt: createdAt(),
    updatedAt: updatedAt(),
  },
  (t) => ({
    pk: primaryKey({ columns: [t.agentId, t.knowledgeBaseId] }),
  }),
);

export const agentsFiles = pgTable(
  'agents_files',
  {
    fileId: text('file_id')
      .notNull()
      .references(() => files.id, { onDelete: 'cascade' }),
    agentId: text('agent_id')
      .notNull()
      .references(() => agents.id, { onDelete: 'cascade' }),
    enabled: boolean('enabled').default(true),
    userId: text('user_id')
      .references(() => users.id, { onDelete: 'cascade' })
      .notNull(),
    createdAt: createdAt(),
    updatedAt: updatedAt(),
  },
  (t) => ({
    pk: primaryKey({ columns: [t.fileId, t.agentId, t.userId] }),
  }),
);
```


这些表展示了 AI 助手（Agent）如何与知识库和文件关联，从而实现更强大的检索和生成能力。

1. `agentsKnowledgeBases` 表

这个表建立了 AI 助手（Agent）和知识库之间的多对多关系。

- `agentId`: 外键，关联到 agents 表，表示特定的 AI 助手
- `knowledgeBaseId`: 外键，关联到 knowledgeBases 表，表示特定的知识库
- `userId`: 外键，关联到 users 表，标识创建该关联的用户
- `enabled`: 布尔值，表示该知识库是否对该 AI 助手启用
- `createdAt` 和 `updatedAt`: 记录关联创建和最后更新的时间

这个表的主键是 `agentId` 和 `knowledgeBaseId` 的组合，确保了每个 AI 助手和知识库的唯一关联。这种设计允许一个 AI 助手可以关联多个知识库，同时一个知识库也可以被多个 AI 助手使用，极大地提高了系统的灵活性和可扩展性。

2. `agentsFiles` 表

这个表建立了 AI 助手（Agent）和文件之间的多对多关系。

- `fileId`: 外键，关联到 files 表，表示特定的文件
- `agentId`: 外键，关联到 agents 表，表示特定的 AI 助手
- `enabled`: 布尔值，表示该文件是否对该 AI 助手启用
- `userId`: 外键，关联到 users 表，标识创建该关联的用户
- `createdAt` 和 `updatedAt`: 记录关联创建和最后更新的时间

这个表的主键是 `fileId`、`agentId` 和 `userId` 的组合，确保了每个用户对每个 AI 助手和文件的唯一关联。这种设计允许 AI 助手关联多个文件，同时一个文件也可以被多个 AI 助手使用，为用户提供了精细的控制粒度。

这两个表共同构成了 RAG 系统中 AI 助手与知识源（知识库和文件）之间的桥梁。通过这些关联，系统可以为每个 AI 助手定制独特的知识背景，从而在进行检索和生成任务时，能够利用最相关的信息来提供更准确、更有针对性的回答。这种设计不仅提高了 AI 助手的性能和适应性，还为用户提供了灵活配置 AI 助手知识范围的能力，是构建高效、个性化 RAG 系统的关键组成部分。

**Reply by arvinxx** [reply]
## File 表

```ts
import {
  boolean,
  integer,
  jsonb,
  pgTable,
  primaryKey,
  text,
  uuid,
  varchar,
} from 'drizzle-orm/pg-core';
import { createInsertSchema } from 'drizzle-zod';

import { idGenerator } from '../../utils/idGenerator';
import { createdAt, updatedAt } from './_helpers';
import { users } from './user';

export const asyncTasks = pgTable('async_tasks', {
  id: uuid('id').defaultRandom().primaryKey(),
  type: text('type'),
  status: text('status'),
  error: jsonb('error'),
  userId: text('user_id')
    .references(() => users.id, { onDelete: 'cascade' })
    .notNull(),
  duration: integer('duration'),
  createdAt: createdAt(),
  updatedAt: updatedAt(),
});

export type NewAsyncTaskItem = typeof asyncTasks.$inferInsert;
export type AsyncTaskSelectItem = typeof asyncTasks.$inferSelect;

export const globalFiles = pgTable('global_files', {
  hashId: varchar('hash_id', { length: 64 }).primaryKey(),
  fileType: varchar('file_type', { length: 255 }).notNull(),
  size: integer('size').notNull(),
  url: text('url').notNull(),
  metadata: jsonb('metadata'),
  createdAt: createdAt(),
});

export type NewGlobalFile = typeof globalFiles.$inferInsert;
export type GlobalFileItem = typeof globalFiles.$inferSelect;

export const files = pgTable('files', {
  id: text('id')
    .$defaultFn(() => idGenerator('files'))
    .primaryKey(),

  userId: text('user_id')
    .references(() => users.id, { onDelete: 'cascade' })
    .notNull(),
  fileType: varchar('file_type', { length: 255 }).notNull(),
  fileHash: varchar('file_hash', { length: 64 }).references(() => globalFiles.hashId, {
    onDelete: 'no action',
  }),
  name: text('name').notNull(),
  size: integer('size').notNull(),
  url: text('url').notNull(),

  metadata: jsonb('metadata'),
  chunkTaskId: uuid('chunk_task_id').references(() => asyncTasks.id, { onDelete: 'set null' }),
  embeddingTaskId: uuid('embedding_task_id').references(() => asyncTasks.id, {
    onDelete: 'set null',
  }),

  createdAt: createdAt(),
  updatedAt: updatedAt(),
});

export type NewFile = typeof files.$inferInsert;
export type FileItem = typeof files.$inferSelect;

export const knowledgeBases = pgTable('knowledge_bases', {
  id: text('id')
    .$defaultFn(() => idGenerator('knowledgeBases'))
    .primaryKey(),

  name: text('name').notNull(),
  description: text('description'),
  avatar: text('avatar'),

  // different types of knowledge bases need to be distinguished
  type: text('type'),
  userId: text('user_id')
    .references(() => users.id, { onDelete: 'cascade' })
    .notNull(),

  isPublic: boolean('is_public').default(false),

  settings: jsonb('settings'),

  createdAt: createdAt(),
  updatedAt: updatedAt(),
});

export const insertKnowledgeBasesSchema = createInsertSchema(knowledgeBases);

export type NewKnowledgeBase = typeof knowledgeBases.$inferInsert;
export type KnowledgeBaseItem = typeof knowledgeBases.$inferSelect;

export const knowledgeBaseFiles = pgTable(
  'knowledge_base_files',
  {
    knowledgeBaseId: text('knowledge_base_id')
      .references(() => knowledgeBases.id, { onDelete: 'cascade' })
      .notNull(),

    fileId: text('file_id')
      .references(() => files.id, { onDelete: 'cascade' })
      .notNull(),

    // userId: text('user_id')
    //   .references(() => users.id, { onDelete: 'cascade' })
    //   .notNull(),

    createdAt: createdAt(),
  },
  (t) => ({
    pk: primaryKey({
      columns: [
        t.knowledgeBaseId,
        t.fileId,
        // t.userId
      ],
    }),
  }),
);
```

1. `asyncTasks` 表

这个表用于跟踪异步任务的执行状态，对于管理文件处理和嵌入生成等长时间运行的任务非常重要。

- `id`: UUID，主键
- `type`: 任务类型
- `status`: 任务状态
- `error`: 存储任务执行过程中的错误信息
- `userId`: 关联到用户表，表示任务的所有者
- `duration`: 任务执行时间
- `createdAt` 和 `updatedAt`: 任务创建和更新时间

2. `globalFiles` 表

这个表存储全局唯一的文件信息，用于去重和共享文件。

- `hashId`: 文件的唯一哈希值，作为主键
- `fileType`: 文件类型
- `size`: 文件大小
- `url`: 文件的访问 URL
- `metadata`: 存储文件的额外元数据
- `createdAt`: 文件创建时间

3. `files` 表

这个表存储用户特定的文件信息，是 RAG 系统中文档管理的核心。

- `id`: 文件唯一标识符
- `userId`: 关联到用户表，表示文件的所有者
- `fileType`: 文件类型
- `fileHash`: 关联到 globalFiles 表，用于文件去重
- `name`: 文件名
- `size`: 文件大小
- `url`: 文件的访问 URL
- `metadata`: 存储文件的额外元数据
- `chunkTaskId` 和 `embeddingTaskId`: 关联到 asyncTasks 表，用于跟踪文件分块和嵌入生成任务
- `createdAt` 和 `updatedAt`: 文件创建和更新时间

4. `knowledgeBases` 表

这个表用于管理知识库，是组织和结构化文档集合的关键。

- `id`: 知识库唯一标识符
- `name`: 知识库名称
- `description`: 知识库描述
- `avatar`: 知识库头像
- `type`: 知识库类型
- `userId`: 关联到用户表，表示知识库的所有者
- `isPublic`: 表示知识库是否公开
- `settings`: 存储知识库的设置信息
- `createdAt` 和 `updatedAt`: 知识库创建和更新时间

5. `knowledgeBaseFiles` 表

这个表建立了知识库和文件之间的多对多关系。

- `knowledgeBaseId`: 关联到 knowledgeBases 表
- `fileId`: 关联到 files 表
- `createdAt`: 关联创建时间

这个表的主键是 `knowledgeBaseId` 和 `fileId` 的组合，确保了知识库和文件之间的唯一关联。

这些表共同构成了 RAG 系统的文档管理和知识库组织框架。`globalFiles` 和 `files` 表管理文件的存储和元数据，`knowledgeBases` 表组织知识结构，`knowledgeBaseFiles` 表将文件关联到特定的知识库。`asyncTasks` 表则支持了文件处理和嵌入生成等异步操作的管理。这种设计不仅支持高效的文档检索和知识管理，还为 RAG 系统提供了强大的可扩展性和灵活性，使系统能够处理大规模的文档集合并支持复杂的知识组织结构。

**Reply by arvinxx** [reply]
## 消息相关

```ts

// if the message contains a file
// save the file id and message id
export const messagesFiles = pgTable(
  'files_to_messages',
  {
    fileId: text('file_id')
      .notNull()
      .references(() => files.id, { onDelete: 'cascade' }),
    messageId: text('message_id')
      .notNull()
      .references(() => messages.id, { onDelete: 'cascade' }),
  },
  (t) => ({
    pk: primaryKey({ columns: [t.fileId, t.messageId] }),
  }),
);

export const messageRAG = pgTable('message_rag', {
  id: text('id')
    .references(() => messages.id, { onDelete: 'cascade' })
    .primaryKey(),
  query: text('query'),
  embeddingsId: uuid('embeddings_id').references(() => embeddings.id, { onDelete: 'set null' }),
});

export const messageFileChunks = pgTable(
  'message_file_chunks',
  {
    id: text('id').references(() => messages.id, { onDelete: 'cascade' }),
    queryId: text('query_id').references(() => messages.id, { onDelete: 'cascade' }),
    chunkId: uuid('chunk_id').references(() => chunks.id, { onDelete: 'cascade' }),
  },
  (t) => ({
    pk: primaryKey({ columns: [t.chunkId, t.id] }),
  }),
);

// convert message content to the chunks
// then we can use message as the RAG source
export const messageChunks = pgTable(
  'message_chunks',
  {
    messageId: text('message_id').references(() => messages.id, { onDelete: 'cascade' }),
    chunkId: uuid('chunk_id').references(() => chunks.id, { onDelete: 'cascade' }),
  },
  (t) => ({
    pk: primaryKey({ columns: [t.chunkId, t.messageId] }),
  }),
);

```

1. `messagesFiles` 表

这个表建立了消息和文件之间的多对多关系。

- `fileId`: 关联到 files 表，表示特定的文件
- `messageId`: 关联到 messages 表，表示特定的消息

主键是 `fileId` 和 `messageId` 的组合，确保了消息和文件之间的唯一关联。这个表允许系统跟踪哪些文件被包含在特定的消息中，对于理解消息上下文和相关资源非常重要。

2. `messageRAG` 表

这个表存储与消息相关的 RAG（检索增强生成）信息。

- `id`: 主键，关联到 messages 表
- `query`: 存储用于检索的查询文本
- `embeddingsId`: 关联到 embeddings 表，表示该消息的嵌入向量

这个表允许系统跟踪每条消息的 RAG 相关信息，包括用于检索的查询和对应的嵌入向量，这对于实现高效的相似性搜索和上下文检索至关重要。

3. `messageFileChunks` 表

这个表建立了消息、查询和文件分块之间的关系。

- `id`: 关联到 messages 表，表示特定的消息
- `queryId`: 关联到 messages 表，表示用于检索的查询消息
- `chunkId`: 关联到 chunks 表，表示检索到的文件分块

主键是 `chunkId` 和 `id` 的组合。这个表允许系统跟踪每条消息检索到的相关文件分块，对于理解消息的知识来源和上下文非常重要。

4. `messageChunks` 表

这个表将消息内容转换为分块，使消息本身也可以作为 RAG 的知识源。

- `messageId`: 关联到 messages 表，表示特定的消息
- `chunkId`: 关联到 chunks 表，表示消息内容的分块

主键是 `chunkId` 和 `messageId` 的组合。这个设计允许系统将消息内容分块存储，使得消息历史也可以作为知识源被检索和利用。

这些表共同构建了一个强大的消息处理和检索框架，是 RAG 系统的核心组成部分：

1. `messagesFiles` 表跟踪消息中包含的文件，为理解消息上下文提供了重要信息。
2. `messageRAG` 表存储每条消息的 RAG 相关信息，支持高效的相似性搜索。
3. `messageFileChunks` 表记录每条消息检索到的相关文件分块，有助于理解消息的知识来源。
4. `messageChunks` 表将消息内容转换为可检索的分块，使对话历史成为有价值的知识源。

这种设计不仅支持从外部文档和知识库中检索信息，还能利用对话历史作为动态知识源。这大大增强了系统的上下文理解能力和信息检索效率，使 RAG 系统能够提供更加智能和连贯的对话体验。

**Reply by zhangleijun111** [reply]
大神什么时候会有一个版本先上线试试啊

**Reply by arvinxx** [reply]
@zhangleijun111 就这周

**Comment by zhangleijun111** [comment]
到时候叫一下如何部署



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: Arvin Xu ***@***.***&gt;
发送时间: 2024年8月19日 20:58
收件人: lobehub/lobe-chat ***@***.***&gt;
抄送: zhangleijun111 ***@***.***&gt;, Mention ***@***.***&gt;
主题: Re: [lobehub/lobe-chat] [RFC] 051 - 知识库/文件上传 (Discussion #3023)

**Reply by arvinxx** [reply]
你的pg实例支持 pgvector 就可以直接部署了。不需要别的东西。

**Reply by zhangleijun111** [reply]
> 你的pg实例支持 pgvector 就可以直接部署了。不需要别的东西。

docker run -p 5432:5432 -d --name pg -e POSTGRES_PASSWORD=mysecretpassword pgvector/pgvector:pg16，这个容器再加上lobechat:database嘛,需要添加什么环境变量嘛？？？

**Reply by arvinxx** [reply]
应该还要一个 APP_URL 指向你的站点域名

**Reply by zhangleijun111** [reply]
> 应该还要一个 APP_URL 指向你的站点域名

到时候出个简易版教程参考下或者，部署失败找一下你问下哈哈哈，直接使用单个lobechat镜像无法部署的是吧

**Reply by SmallAi-API** [reply]
> 你的pg实例支持 pgvector 就可以直接部署了。不需要别的东西。

请问是直接拉取“[feat/file-and-knowledge-base](https://github.com/lobehub/lobe-chat/tree/feat/file-and-knowledge-base)”分支吗

**Reply by arvinxx** [reply]
@Xiaokai6880 对

**Reply by zhangleijun111** [reply]
> 应该还要一个 APP_URL 指向你的站点域名

还没有发布急死了

**Comment by cmj2002** [comment]
请问能支持在文件管理界面中的图像预览吗？目前删除对话之后不会删除对话中的图片，没有预览的话在文件管理页面查找和删除比较麻烦。

**Reply by arvinxx** [reply]
已经支持

**Comment by zwjzxh520** [comment]
请问文件分片时，可以使用国产大模型的API吗？而不是像现在这样固定使用 openai 的

**Comment by Snowflake-Pink** [comment]
![image](https://github.com/user-attachments/assets/0af82566-93a8-488d-9d77-d0b2d2d94b30)
想请教一下，我找了官网文档和README各个地方，似乎都没有看到如何配置 Lobechat 对接使用 Unstructured.io 的配置指引。
![image](https://github.com/user-attachments/assets/d51d103b-2bb5-4aff-8b62-8f5903f8f227)
从上图来看，似乎是配置 `UNSTRUCTURED_API_KEY` 和 `UNSTRUCTURED_SERVER_URL` ？是否还需要设置另外一个值来使 Lobechat 的分块任务指向 unstructured ？

---------
![image](https://github.com/user-attachments/assets/0dd3207e-592b-48d9-8687-5f321899b8b7)
另外，似乎源码中似乎还有 doc2x 的配置可选，这一文档处理方式是否已经成熟可选？

谢谢♥️
